supplemental_notebooks/2021-09-20_muddiest_points.Rmd

```diff
@@ -0,0 +1,176 @@
+```
+
+`BiocManager::install()` tends to be a bit more chatty than the base R installation; be on the lookout for prompts that you need to respond to in the console.
+
+## Logical vectors
+
+I briefly talked about logical (Boolean) vectors during the first session, but they can definitely be confusing at times, so here I will go into a bit more depth. 
+
+First off, what is a logical vector? A logical vector is one that contains only the values `TRUE` and `FALSE` (and potentially `NA`).
+Notice the all caps: R requires this. 
+
+So the simplest way to create a logical vector is just like another vector, using `c()` but only those values:
+
+```{r}
+my_bool <- c(TRUE, FALSE, FALSE, TRUE)
+my_bool
+class(my_bool)
+```
+(Full disclosure, you _can_ use `T` and `F` in addition to `TRUE` and `FALSE`, but this is a bad idea, because it is possible to overwrite those abbreviations like normal variable names. `F <- TRUE` is a perfectly valid statement, even though it can wreak havoc on sloppy code.)
+
+
+That isn't too useful, generally, because we usually want to create our logical vectors based on some data. To that we will tend to use one of a few comparison or logical operators:
+
+Starting wiht the comparisons:
+* `==`: equal to
+* `<`: less than
+* `<=`:  less than or equal to 
+* `>`: greater than
+* `>=`: greater than or equal to
+* `%in%` : is a member of some other vector
+
+Let's say we have a vector of (semi) random numbers. 
+
+```{r}
+rand_nums <- c(0.74, -0.83, 0.42, 0.29, -0.825, 0.25, 1.68)
+```
+
+We can use comparisons to create new logical vectors:
+
+```{r}
+negative <- rand_nums < 0
+negative
+```
+
+Or we can test which of some character vectors match a particular value:
+```{r}
+words <- c("hello", "goodbye", "howdy", "aloha", "cat")
+is_cat <- words == "cat"
+is_cat
+```
+
+or are in a set of known values: 
+```{r}
+greetings <- c("aloha", "hello", "hi", "hola", "howdy")
+is_greeting <- words %in% greetings
+is_greeting
+```
+
+We can also modify combine multiple boolean vectors with logical statements to get new boolean vectors.
+
+* `!`: not (inverts `TRUE` and `FALSE` )
+* `|` : logical **or** - if either of the elements are `TRUE`, this gives `TRUE`
+* `&` : logical **and** - if both of the elements are `TRUE`, this gives `TRUE`
+
+These logical operators will act on each of the elements of a vector, or each pair from two vectors (note that I am not saving these to new vectors, but you could!):
+
+```{r}
+!is_greeting
+```
+
+```{r}
+# Is it a greeting or is it a cat, because either way, we should say hi back
+is_greeting | is_cat
+```
+
+
+Once we have a logical vector with the values we are after, we can use them in selections, much like we might use the position numbers, to get only those values that gave us `TRUE` in whatever comparison we were doing:
+
+```{r}
+greet_words <- words[is_greeting]
+greet_words
+```
+
+We can also use them in many other places, such as the `dplyr::filter()` statements we covered in lecture.
```

Comment:
Should we point out the notebook/an example in the notebook where we implemented `dplyr::filter()` here -- I am thinking it will be easier to connect the dots for a participant following along here but also don't feel so strongly about it!